server/job.go

```diff
@@ -0,0 +1,114 @@
+package main
+
+import (
+	"github.com/google/uuid"
+	pb "github.com/romanlevin/felek/jobs"
+	"log"
+	"os/exec"
+	"sync"
+)
+
+type job struct {
+	id        string    // A UUID string
+	cmd       *exec.Cmd // The underlying Cmd object wrapping the job process
+	stopped   bool      // Has this job been stopped by a client?
+	exitError error     // The error returned by Wait on the Cmd
+	owner     string    // The username of the user who started the job
+	lock      *sync.RWMutex
+}
+
+// exitedWithExitCode assumes a lock is held
+func (j *job) exitedWithExitCode() bool {
+	return j.exitError == nil && j.cmd.ProcessState != nil
+}
+
+// running assumes a lock is held
+func (j *job) running() bool {
+	return j.exitError == nil && j.cmd.ProcessState == nil
+}
+
+func (j *job) stop() error {
```

Comment:
What happens if multiple clients try to stop the job simultaneously?

Reply:
One would get the lock and kill it, the others would error out trying to kill a killed process, or succeed in sending the signal if the process is still alive by that time.

I could 'Wait()' here to make sure the process is dead when the lock is released, but then it might complicate things with 'wait', not sure what happens with multiple waiters.

Reply:
Turns out calling `.Wait()` on a finished processes returns an error (`"exec: Wait was already called"`), but in any case, I think `Stop` returning an error for all but one of multiple clients is working as intended.

Reply:
Any way to handle this more gracefully and prevent clients besides the first one try to kill the job?

Reply:
Yeah, I brought it up earlier in this thread. We can `Wait` in `Stop`, before releasing the lock.

Another option is to check `stopped` before calling `Kill`.

Reply:
I implemented the latter.